### ISSUE TITLE
Task View (or similar) functionality

### ISSUE THREAD

**Issue by JulienNic0las** [solus-project/budgie-desktop#334]
Pretty sure that this is not the right place for such a request, but I give it a try. 
I found nowhere the possibility to display currently opened windows/applications similarly to the "task view" of gnome shell (or other popular DE, even the recent W10) via a keyboard shortcut. I think this is a great missing feature.

Bye

**Comment by JoshStrobl** [comment]
Do you mean the Expose feature in OS X or the Window Spread feature in Shell?

Agreed that it is a "must-have", I prefer it over traditional alt+tab. Sounds like something that could be implemented in budgie-wm.

**Comment by JulienNic0las** [comment]
Yes excatly, but maybe less messy than the examples I found on google image with "Expose feature mac os" as keywords... 

Alt-tab is great but, in my opinion, not usable with a lot of opened windows.

**Comment by starquake** [comment]
I too would prefer an exposê-like or mission control like solution over a Windows like solution.

Here's a screenshot of how Mission Control currently looks like in the latest version: http://imgur.com/SLMAd48

There's also an option to group the Windows by application:
http://imgur.com/9Fj3Chq

I agree it looks messy but it's really easy to use. I almost always instantly see which Window to switch to.

**Comment by Redneckita** [comment]
That's a feature that i would really like to have!

**Comment by marksev1** [comment]
+1

**Comment by ikeydoherty** [comment]
Please people **STOP** necromancing issues. It's open - it'll be dealt with

**Comment by openAfre** [comment]
Coming from elementary and gnome-shell I am missing these :"mission overview" modes quite a bit.

Something like:
All windows all Desktops -> Super
All windows current Desktop -> Super-W
All applications from current app type (all chrome windows) -> Super-A

Also moving windows between virtual desktops like gnome shell or mission overview elementary style would be just marvelous!

**Comment by marcosfede** [comment]
any updates on this?

**Comment by JoshStrobl** [comment]
@marcosfede As indicated by the milestone, this isn't something that'd be addressed until Budgie 11, which we haven't started development of yet (we're readying for a 10.4.1).

**Comment by mfed3** [comment]
Very useful, even Windows 10 has this feature and it works very nicely with multi monitor support. Emulating the functionality there should be the goal, not worrying about fancy animations like expose has.

**Comment by M4gicT0** [comment]
It's the only feature I really miss from Budgie...

**Comment by ripefig** [comment]
I'd recommend copying Gnome overview. No reason why you shouldn't be able to search, launch and manage workspaces from the the overview.
Also overlay icons are a must, especially for people with poor eyesight. 

Prior art: 

Elementary
![image](https://user-images.githubusercontent.com/45201036/55760776-c718d180-5a65-11e9-813b-ee564f668a05.png)

Gnome overview (just get rid of the gray background and make the icon highly transparent - I've used it and you can see the window content perfectly well under the transparent icon):
![image](https://user-images.githubusercontent.com/45201036/55760820-ed3e7180-5a65-11e9-8172-8270d38ef35d.png)

**Comment by lvanderwiele** [comment]
@ripefig It may be a little late but I disagree with you on two points.

1. First, Overview is a core functionality of GNOME (it's even the centre of GNOME Shell) and it would be useless to copy it as it doesn't offer something else to Linux desktops. Budgie should try to be its own thing (at least as far as Linux desktops go).
2. Although I can see why some people would want those icons, I personally don't like them. Since I often work with multiple windows of the same apps: multiple terminals, multiple Firefox windows (often open with either Wikipedia, documentation, Moodle or something similar), multiple PDFs opened in Evince (all of which are formatted with LaTeX and thus similar) or a collection of MATLAB plot. If each time I start a MATLAB script I end up having 5 or 6 plot windows, trying to differentiate them under the MATLAB icon won't be easy. Should this be implemented, I'd really like to have a way to mask them.

Personally, I think we should go with something similar to Mission Control but which doesn't mask the panel (i.e. if I have a top panel I don't want it to be masked by the virtual desktop's "Dock"). Additionally, I could suggest a few feature

* In Unity 7, if multiple windows of an application existed, clicking on its icon would have started a window spread comprising only the windows of that app. A second click would then miniaturize all windows of that application. A similar feature could be useful
* Still in Unity 7, when in Window spread, one could start typing causing a hidden search field to appear and the shown window would be sorted according to the search the user started. For example, if I hit Super+W (to enter window spread) and started typing "firefox", Unity would have hidden all windows not comprising "firefox" in the title.
* To stop drawing inspiration from Unity and start from macOS, I think we could implement a way for maximized windows to move to a dedicated virtual desktop. In this desktop the panels would adopt a autohide mode to give as much space as possible to the window (but would still be available by pressing on the border of the screen). If possible, hiding the titlebar of non-CSD apps could also be an idea. Typically, I often work with a full-screen Atom or a fullscreen Firefox when using Overleaf and having those automatically move to a dedicated desktop could be nice. However, only the windows maximized using the shortcut (Super+Up) or the maximize setting would be affected. Double-clicking the titlebar or dragging the window to the top of the screen would any cause the window to resize itself to the size of the screen (i.e. current Windows-style maximization).
* Similarly, I think we can draw inspiration from Windows 10 for the tiling behaviour. When tiling a windows on Win10, the remaining windows of the desktop are spread next to it, allowing the user to automatically tile another one on the remaining part of the screen. Additionally, we could consider the two newly tiled windows as a full screen one and move the couple to a dedicated desktop. Should one of the two windows be closed (or simply resized to its previous size (by hitting Super+Down for example)), the remaining window(s) would be resized to its (their) previous size(s) and returned to its (their) original desktop.

I don't know how much of that can be implemented for Budgie 11 (or simply can be implemented at all) but i think these would be great additions to the desktop.

**Comment by ripefig** [comment]
> In Unity 7, if multiple windows of an application existed, clicking on its icon would have started a window spread comprising only the windows of that app. A second click would then miniaturize all windows of that application.

I disagree. This requires you to do too much mousing, since you have to move the cursor to the dock and then move to the center of the screen. If you going to have a dock, the quickest way to switch windows is to just keep clicking on the same group icon until you get the right window. The second quickest way, is to have window previews directly over the dock, like on Windows.


> Since I often work with multiple windows of the same apps

You could group windows of the same app and have just one icon for the entire group. Just take this design from IOS and pretend that the workspace group are actually windows of the same app.

![image](https://user-images.githubusercontent.com/45201036/59163391-3ca62a00-8b09-11e9-95f1-443165db5aed.png)

Of course it would be ideal to have an API that applications could use to draw their own previews. It's pie in the sky for most programs, but it could actually be implementable for a few critical programs like Firefox with an extension (and that's really half the job - most of your windows will be browser windows/tabs). 

I think the overview is better than any other solution because it allows you to do literally anything, in one interface. There are many possibilities for the overview that people haven't yet thought of. If you just go with the dock, you'll be locking yourself out of all that potential.

side note: I am not sure how much Budgie can really do without full time developers. But if they have an attractive list of planned features and a budget, they might be able to raise enough money on Kickstarter for full time development.

**Comment by JoshStrobl** [comment]
An exposé feature similar to GNOME Shell is planned for Budgie 11 and no earlier. The intent of the feature is to provide functionality in that regard which is similar to GNOME Shell to enable those individuals to have that experience without the performance reductions you'd otherwise see with GNOME Shell (and its use of JavaScript + GJS).

> First, Overview is a core functionality of GNOME (it's even the centre of GNOME Shell) and it would be useless to copy it as it doesn't offer something else to Linux desktops. Budgie should try to be its own thing (at least as far as Linux desktops go).

Who cares if it doesn't "offer" something else to Linux desktops? The whole point is you can have it all **under one desktop**. You get to have the **optional, opt-in** exposé mode, Raven, and other changes I plan on bringing in Budgie 11, all under one roof. If you're arguing that we should do our own thing then:

1. We shouldn't have Budgie Menu...because other desktops have menus, right?
2. We shouldn't have Icon Tasklist because well, other desktop environments have icons which you can use to select an application.
3. We shouldn't have the traditional Taskview because other desktop environments have those, such as MATE.
4. We shouldn't have alt+tab because other desktop environments have that ability to switch between windows.

I can go on...

**Comment by lvanderwiele** [comment]
@ripefig 
> I disagree. This requires you to do too much mousing, since you have to move the cursor to the dock and then move to the center of the screen. If you going to have a dock, the quickest way to switch windows is to just keep clicking on the same group icon until you get the right window. The second quickest way, is to have window previews directly over the dock, like on Windows.

Not necessarily. For example, if you map Super+1/2/3/4/... shortcuts to the icons in your task list, you could simply type Super+2 to spread all windows of the second app and then select the window you want with arrow keys or your mouse. It's simply an Exposé feature that discriminate between windows.

Personally, I don't like having the window previews above the icons because I don't find it very legible. Again, if I have multiple terminals, Firefox windows with similar sites or evince windows with similar documents (which happens a lot), it gets out of hand really fast.

> You could group windows of the same app and have just one icon for the entire group. Just take this design from IOS and pretend that the workspace group are actually windows of the same app.

macOS does that. As was shared above:

![](https://i.imgur.com/9Fj3Chq.jpg)

As you can see, everything is grouped by applications and the groups spread when hovered upon with the mouse. It's also a behaviour that could be implemented. That being said, the smaller offcentered  icons make it more legible than what you linked in your previous post.

@JoshStrobl 

> An exposé feature similar to GNOME Shell is planned for Budgie 11 and no earlier. The intent of the feature is to provide functionality in that regard which is similar to GNOME Shell to enable those individuals to have that experience without the performance reductions you'd otherwise see with GNOME Shell (and its use of JavaScript + GJS).

Which is why I'm making these suggestions on the way windows, virtual desktops and exposé could work in Budgie 11.

> Who cares if it doesn't "offer" something else to Linux desktops? The whole point is you can have it all under one desktop. You get to have the optional, opt-in exposé mode, Raven, and other changes I plan on bringing in Budgie 11, all under one roof. If you're arguing that we should do our own thing then:
> * We shouldn't have Budgie Menu...because other desktops have menus, right?
> * We shouldn't have Icon Tasklist because well, other desktop environments have icons which you can use to select an application.
> *  We shouldn't have the traditional Taskview because other desktop environments have those, such as MATE.
> *  We shouldn't have alt+tab because other desktop environments have that ability to switch between windows.
>
> I can go on...

The thing is not that we can't have an Exposé/Mission Control because someone else did it. It's that Overview, as it exist in GNOME Shell is way more than: it is the core of GNOME's desktop experience. It's an Exposé mode but also the only way to open application or browse installed application as well as a desktop search function allowing to search for files, applications, calendar entries, available software inside GNOME Software, etc.

I think we should have a mission control feature but I don't think we should integrate everything in it, like GNOME does.

We can use menus, tasklists, alt-tab, etc. because everyone has one or the other implementation of it. But everyone also creates its own desktop experience, bringing one feature or the other to the table. Budgie, for example, brings a solid desktop-focused experience that offers a level of customizability inspired by KDE (with the applets) but using GTK and GTK apps and adds Raven to act as a widget/notification centre (or at least, that's my vision of Budgie).
